python/cuml/cuml/internals/type_reflection.py

```diff
@@ -0,0 +1,237 @@
+# Copyright (c) 2025, NVIDIA CORPORATION.
+
+from cuml.internals.global_settings import GlobalSettings
+from cuml.internals.array import CumlArray
+from cuml.internals.memory_utils import using_output_type
+from cuml.internals.input_utils import (
+    determine_array_type,
+    determine_array_dtype,
+)
+
+import inspect
+from collections.abc import Sequence
+from contextlib import contextmanager
+from dataclasses import dataclass
+from functools import wraps
+from typing import Any
+
+import cupy as cp
+import numpy as np
+
+# Everything related to API boundary determination
+
+
+def in_internal_api() -> bool:
+    return GlobalSettings().api_depth > 1
+
+
+def cuml_public_api(func):
+    @wraps(func)
+    def inner(*args, **kwargs):
+        GlobalSettings().increment_api_depth()
+        try:
+            return func(*args, **kwargs)
+        finally:
+            GlobalSettings().decrement_api_depth()
+
+    return inner
+
+
+# CumlArray
+
+
+def as_cuml_array(X, dtype=None) -> CumlArray:
+    """Wraps array X in CumlArray container."""
+    # TODO: After debugging, replace this with an immediate call to the CumlArray constructor.
+    return CumlArray(X, dtype=dtype)
+
+
+# CumlArrayDescriptor
+# TODO: Replace CumlArrayDescriptor in cuml/common/array_descriptor.py
+
+
+class CumlArrayDescriptor:
```

Comment:
This will be moved into the `array_descriptor` module once the class is finalized.